SocketRocket/Internal/Security/SRSecurityPolicyBuilder.m

```diff
@@ -0,0 +1,213 @@
+//
+// Copyright (c) 2016-present, Facebook, Inc.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree. An additional grant
+// of patent rights can be found in the PATENTS file in the same directory.
+//
+
+#import "SRSecurityPolicyBuilder.h"
+
+#import "SRLog.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ `SRPinningPolicy` requires some level of encryption and allows pinning to
+ provider certificates.
+ */
+@interface SRPinningSecurityPolicy : NSObject <SRSecurityPolicy>
+
+@property (nonatomic, strong, readonly) NSArray *pinnedCertificates;
+
+@end
+
+@implementation SRPinningSecurityPolicy
+
+- (instancetype)initWithPinnedCertificates:(NSArray *)certificates
+{
+    self = [super init];
+    if (!self) return self;
+
+    if (certificates.count == 0) {
+        @throw [NSException exceptionWithName:@"Creating security policy failed."
+                                       reason:@"Must specify at least one certificate when creating a pinning policy."
+                                     userInfo:nil];
+    }
+    _pinnedCertificates = [certificates copy];
+    return self;
+}
+
+- (void)updateSecurityOptionsInStream :(NSStream *)stream
+{
+    // Enforce TLS 1.2
+    [stream setProperty:(__bridge id)CFSTR("kCFStreamSocketSecurityLevelTLSv1_2") forKey:(__bridge id)kCFStreamPropertySocketSecurityLevel];
+
+    // Do not validate certificate chain for the stream since we're pinning.
+    NSDictionary<NSString *, id> *sslOptions = @{ (__bridge NSString *)kCFStreamSSLValidatesCertificateChain : @(NO) };
+    [stream setProperty:sslOptions forKey:(__bridge NSString *)kCFStreamPropertySSLSettings];
+}
+
+- (BOOL)evaluateServerTrust:(SecTrustRef)serverTrust
+                  forDomain:(NSString*)domain
+{
+    SRDebugLog(@"Pinned cert count: %d", self.pinnedCertificates.count);
+    NSUInteger requiredCertCount = self.pinnedCertificates.count;
+    if (requiredCertCount == 0) {
+        @throw [NSException exceptionWithName:@"Evaluating security policy failed."
```

Comment:
Don't think this is required, since we validate the number of pinned certificates on creation of the policy.